### ISSUE TITLE
How to model Service

### ISSUE THREAD

**Issue by fosterlynn** [valueflows/valueflows#173]
We have discussed this off and on in different contexts, but I can't find an issue dedicated to it, so I'm starting one.  This is triggered by the work done on exchange, here: https://github.com/valueflows/exchange/issues/32.  The transfer model (https://github.com/valueflows/exchange/issues/32#issuecomment-267834962) makes it a lot easier to model Service in a simpler way that also more corresponds to reality, at least when the transfer of the service needs to be recorded. 

This relates to both Process and Exchange, providing a service, and 'transferring' that service to one or more agents.  Treating a service like a material resource leads to clunkiness - you create some kind of "service resource", which then is transferred to another agent(s).  The process side is pretty clean, since services tend to involve at least a work input and often more than that.  But in fact once the process is complete the service has been transferred, like it or not.  Once someone's hair is cut, it is cut.

The exchange model we have been working on uses events rather than resources.  This gives freedom to for example exchange work for something else.  It also in similar manner should give the freedom to exchange a service without creating a somewhat mythical resource to exchange.

Proposal:  We create a new vf:action verb (type of event) for service outputs, called something like provideService, corresponding to vf:produce for material items, i.e. the output of the process that provides the service. The vf:affectedResource referenced by the event will just have information on the model and/or category, there will be no instantiated vf:Resource.

(Note this is separate from the question of does the person's hair or whatever go into and out of the process, like an automobile that is being repaired might do, which I'm not really addressing here.)

When a service will be exchanged for something else, the new exchange model works well. The output provideService event will create a commitment for the reciprocal event (like payment for the haircut).

The problem we ran into when discussing this (https://github.com/valueflows/exchange/issues/32#issue-193362583) is what do we do when a service is output of a process, and also input into another process, rather than being exchanged.  The example is when Elf Consulting Services provides a service of participating in a ValueFlows meeting.  Elf Consulting Services has a process where Pavlik does some work, uses a computer and internet, and outputs the service.  The ValueFLows process of conducting the meeting wants to input that service.  The output and the input are the same thing.  Proposal: We use the same event as output of one process and input to another process. This doesn't follow the basic value flow pattern of I > P > O > R > I > P > O.  So that seems unfortunate.  But it also seems like the best representation of reality.  Something like I > P > O/I > P > O.  And should be easy to track in a chain, even though it is an exception.

Are we close to finalizing the "service" question?  Or not yet?

I can write some yaml.

**Comment by bhaugen** [comment]
Let's list enough representative examples of services for test coverage:
* Repair or improvement:
    * haircut
    * medical or dental treatment
    * auto repair
    * lawnmowing or snow shoveling
    * plumbing service
* Care work
    * baby sitting
    * pet sitting
    * invalid or elder care
* Consulting service
* Contract programming

What else?

**Comment by bhaugen** [comment]
When the output of a service process becomes an input to another process: this is the same structurally as a situation we encountered in Sensorica where an intangible resource was an output of one process and an input to the next. This is also the same as @elf-pavlik 's "indirectly identified resources". 

In Sensorica, we never found a good way to chain such value flows together, because the code assumed an output resource that went into the next process, but the resource did not exist in any way that we could get a handle on. So it was always a kludgy exception. Lynn's proposal upthread would handle it. It's also an exception, but I think a clean one.

**Comment by elf-pavlik** [comment]
>  where an intangible resource was an output of one process and an input to the next. This is also the same as @elf-pavlik 's "indirectly identified resources".

intangible doesn't imply "indirectly identified" so I wouldn't make that connection here

> Proposal: We use the same event as output of one process and input to another process. This doesn't follow the basic value flow pattern of I > P > O > R > I > P > O. So that seems unfortunate. But it also seems like the best representation of reality. Something like I > P > O/I > P > O. And should be easy to track in a chain, even though it is an exception.

In case where service output by one process becomes input to another process, we have two distinct processes which in a way make together a bigger process. Producing and consuming the service happens at the same time. Which makes it in a way act as bridge between processes. We can see how things look if we don't create *ServiceResource* (which seems like serial one). We can't just evaluate it by looking at *Observation* layer but also should at least include *Plan* (Intent + Commitment). I think we should model work and service in similar way, if we want to model *work* without resource, we also need a way to have it as output (no magical inputs that come out of nowhere, whatever process(es) produce work also has inputs!).

> Let's list enough representative examples of services for test coverage:

* Transportation
  * Freight
  * Passenger

We should look at all multiple variants
* single transportation process outputs multiple services
* multiple transportation processes outputs add up to a single service (possibly represented as bundle of small services - which fits pattern where multiple commitments can satisfy single intent and multiple events can fulfill single commitment)

* Multiple batches of apple juice - we discussed possibility of having
  * full day process which has
    * input: use - apple juice press owned by Judy
    * input: work - Dan doing it as one time gig
    * input: work - coordination/logistics by Melanie
    * output: service x 3 - for each juice batch
 * 3 processes per batch of apples
  * input: consume - apples owned by agent X|Y|Z
  * output: produce - apple juice owned by agent X|Y|Z

**Comment by bhaugen** [comment]
> intangible doesn't imply "indirectly identified" so I wouldn't make that connection here

Then, substitute "impossible or impractical to directly identify".

Apples are directly identifiable.

**Comment by elf-pavlik** [comment]
Many services also seem directly identifiable and in many ways look similar to 'serial resource'.

> The example is when Elf Consulting Services provides a service of participating in a ValueFlows meeting. Elf Consulting Services has a process where Pavlik does some work, uses a computer and internet, and outputs the service. The ValueFLows process of conducting the meeting wants to input that service.

Case above I would see as possible and practical to directly identify. If represented by a resource it would fits closer resource 'managed as serial' than resource 'managed as stock'. If represented by event i also don't see any problem with directly identifying that event.

**Comment by bhaugen** [comment]
How would you identify "Elf Consulting Services provides a service of participating in a ValueFlows meeting"?

**Comment by fosterlynn** [comment]
>Many services also seem directly identifiable and in many ways look similar to 'serial resource'.

Yes if we do create a resource, they do seem like 'serial resources' in that you always need the exact one, you can't just take from a pile of substitutable services.

I did this in NRP for Sensorica for services around 3d printer use, but I didn't like it, it was very clunky, and I ended up with resources hanging around that were meaningless.

That said, it did work.  So we have a choice.

But also thinking about exchanging services vs using services in a process - it would be good it we did those the same way.  Our latest exchange model uses the events so doesn't require creation of a resource - it solves the service question nicely when the service is given as part of an exchange.  So if we can figure out the cleanest way of not instantiating a service resource when it goes into the next process, that would be nice.

>We can't just evaluate it by looking at Observation layer but also should at least include Plan (Intent + Commitment). 

I don't understand this, why would we need a plan, if there wasn't a plan in real life?  Either method (using the same event, or creating a serial-like resource that goes into the next process) should work just on the observation layer.  What am I missing?

**Comment by fosterlynn** [comment]
>I think we should model work and service in similar way, if we want to model work without resource, we also need a way to have it as output (no magical inputs that come out of nowhere, whatever process(es) produce work also has inputs!).

If someone wants to model work this way, fine.  I see some challenges though, like how much of what you ate today goes into this particular work output, and how much goes into that work output, and how much goes into your free time.  Etc.  I do see the value in the aggregate of knowing what is needed / wanted for people's ongoing health and happiness and ability to contribute.  In fact it will be essential to have a more rational way to know what we should be producing for people's ongoing lives.  But to me that will happen on a different level than each work event.  But that's just my view, and I think VF should support such an activity.

I think for myself, since we don't have an immediate request for that use case, I would like to focus on service, and then see if that pattern can apply to work as you would like to do.

**Comment by fosterlynn** [comment]
>When the output of a service process becomes an input to another process: this is the same structurally as a situation we encountered in Sensorica where an intangible resource was an output of one process and an input to the next.

@bhaugen I think you are talking about resources that were not instantiated because it was not worth having them in the system as inventory, but which were touchable in real life after they were produced or purchased.  I think this has similarities, but is different than the service structure, because of @elf-pavlik 's observation about services needing to be serial-like, i.e. individually identified.  I don't think the pattern of using the same actual event as output of one process and input to another process would work for the non-inventoried but material resources, since many of those are substitutable and managed as stock.  We should at some point look at this situation and see if the information we are carrying about this type of uninstantiated resource is enough for the value chain processing to work cleanly.

But it does give me an idea for another option, which is instead of using the same event as output and input when it is process to process, we could use different events with unique identifying information about the service on both events, but still not instantiate the resource.  Not sure of advantages or disadvantages to this, except that in fact the event is in real life the same.

**Comment by bhaugen** [comment]
> @bhaugen I think you are talking about resources that were not instantiated because it was not worth having them in the system as inventory, but which were touchable in real life after they were produced or purchased.

No, it was concepts like "Conditions for Project" and "Mapping Interest"  as seen in this recipe:
http://nrp.sensorica.co/accounting/view-stream-recipe/329/

**Comment by bhaugen** [comment]
> I  do see the value in the aggregate of knowing what is needed / wanted for people's ongoing health and happiness and ability to contribute. In fact it will be essential to have a more rational way to know what we should be producing for people's ongoing lives. 

The Mutual Aid Networks want to do this. They call it Full Cost Accounting. I committed to help them. Haven't gotten started yet. When they do, I will open a new VF issue to discuss it.

**Comment by stevebosserman** [comment]
For what it's worth, in our timebank we have identified five major
categories of services our members can do for one another:
1) Companionship (face-to-face and virtual);
2) Advocacy (includes interfacing with various nonprofits, government
agencies, businesses, medical practitioners, financial institutions, etc.
on behalf of members as well as engagement in governance / political
processes);
3) Distribution (transportation for a member or pick-up and delivery of
goods on behalf of a member);
4) Handiness (repairs, assistance with moving / rearranging furniture,
housecleaning / home organizing, etc., etc.); and,
5) Professional (something that often utilizes specialized skills,
equipment, processes, etc. and often requires certification / licensure on
the part of those delivering the service

These are ordered from least (1) to most (5) in terms of complexity /
rigor.  Timebankers adhere to the credo that everyone, unless they are
incapacitated, has something of value to offer.  According to our
categories, that means at a minimum anyone can offer (1) companionship to
another.  Then, depending on wherewithal folks can offer increasingly
demanding services.

I might add that as we get our heads around what it means to be able to
sustainably meet our basic needs and govern accordingly in a
community-as-a-node-in-a-global-network mode, we will see a substantial
uptick in time spent by community members in doing companionship and
advocacy.  The more the design of value exchange systems (like what you're
working on!) reflect the importance and significance of these types of
activities, the faster community-based social / political / economic
systems will change.  And that's a VERY GOOD THING!  So, THANKS for doing
what you do!

**Comment by bhaugen** [comment]
Nice categories, @stevebosserman !

**Comment by elf-pavlik** [comment]
> If someone wants to model work this way, fine. I see some challenges though, like how much of what you ate today goes into this particular work output, and how much goes into that work output, and how much goes into your free time. Etc.

If we compare it to transportation process which provides many services of transporting freight or passengers. We can say how much fuel etc. we can attribute to transporting of each thing. Still we have a way to track that certain amount of fuel got consumed in a process of providing particular services.

I'd like to avoid having *work* as magic input which doesn't come as output of some other process(es) which in turn may seem like it doesn't require any inputs for creating that output (work).

I think I would see *work* as a special case of *service* and possibly by figuring out service we would have model for work by just adding few extra constraints.

**Comment by bhaugen** [comment]
I agree it is important to include all of the resources and processes required for daily life in a complete economic network model. Likewise all of the rest of the surrounding ecosystem. 

I think work is different from some other things that would be called services, though, and think it deserves its own issue, which I will start now. See #174
